accepted/00000-ansible-integration.md

```diff
@@ -0,0 +1,327 @@
+
+Notice also that the Ansible control node is not listed as systems list page, since there is not system entry for it, it would be shown in the respective new UI page for managing Ansible control nodes.
+
+#### New UI tabs & pages
+
+* Ansible
+* Ansible / Ansible control nodes list
+* Ansible / Add new Ansible control node
+* Ansible / View Ansible control node information & and visualize playbooks (would also allow triggering playbooks via "salt-ssh" in the control node)
+* Systems / Systems list / Ansible (show only Ansible systems and its control node)
+
+or even create a new entry level on the menu as "Automation", something like:
+
+* Automation / Ansible / ...
+* Automation / Salt / ...
+* Automation / Remote commands
+
+## Operating Ansible
+
+This part describes different expectations/features to implement in order to execute certain operation in your Ansible infrastructure.
+
+#### Run playbooks in an Ansible control node node
+Each Ansible control node node contains an Ansible inventory together with all different playbooks and files used together with the playbooks. The targets for the different tasks of a playbook are defined inside the playbook yaml file itself and not externally like Salt does for the states.
+
+This is an example of Ansible playbook:
+
+```yaml
+---
+- name: update web servers
+  hosts: webservers
+  remote_user: root
+
+  tasks:
+  - name: ensure apache is at the latest version
+    yum:
+      name: httpd
+      state: latest
+  - name: write the apache config file
+    template:
+      src: /srv/httpd.j2
+      dest: /etc/httpd.conf
+
+- name: update db servers
+  hosts: databases
+  remote_user: root
+
+  tasks:
+  - name: ensure postgresql is at the latest version
+    yum:
+      name: postgresql
+      state: latest
+  - name: ensure that postgresql is started
+    service:
+      name: postgresql
+      state: started
+```
+
+As you can see, this playbook is defining different tasks to execute on different "hosts" groups. Those groups are of course defined in the Ansible inventory (default at /etc/ansible/hosts).
+
+The execution of a given playbooks is done in the corresponding Ansible control node, which contains the inventory and the hosts definition.
+
+The playbooks available to apply on each Ansible control node, are the ones exposed by "ansible-gatherer" and ultimately, in the future, playbooks created and maintained in Uyuni (via future Ansible playbook catalog) that would be then pushed to the control node node.
+
+At the time of running a playbook from the Ansible control node, Uyuni is able to do it by reaching the control node and operating it using "salt-ssh" when it's an external host. If the control node is the Uyuni server or a registered minion we don't necessarily need to use "salt-ssh".
```

Comment:
Should our main goal just be to make all these systems regular minions and avoid using salt-ssh when we can?
I think it is nice that it would work also with salt-ssh, but as it has massive performance problems we better design everything to become real minions and leave salt-ssh as a corner case for special situations.
What do you think?

Reply:
> Should our main goal just be to make all these systems regular minions and avoid using salt-ssh when we can?

Yes, that's what we agreed on when we discussed the RFC a few weeks ago.

Also, using minions would allow us to avoid user+password. 

> I think it is nice that it would work also with salt-ssh, but as it has massive performance problems we better design everything to become real minions and leave salt-ssh as a corner case for special situations.
> What do you think?

I don't think performance is a factor here, though.

It's the Ansible control nodes that will be registered to Uyuni. That's a handful of nodes, tops. Then each of those Ansible control nodes manages (always via SSH) the clients (third level of depth: 
Uyuni Server (salt-master) => Ansible control node (salt-minion) => Client (ansible)

In fact, as a curiosity (it does not affect this RFC for anything), if you think about it the final design of this Ansible integration is not that different from the Hub, where we also have three levels:
Uyuni Hub Server (salt-master) => Uyuni Server (salt-master and salt-minion) => Client (salt-minion)

Reply:
Seems here is a big missunderstanding. I thought we Salt/Ansible is for every client the goal. It is a salt client in Uyuni with the additional capability to execute also Ansible Playbooks. In this case I would say salt-ssh is a performance problem.

Reply:
I am now confused myself and need to re-read the RFC %)

Reply:
@mcalmer the Uyuni Server will contact, via SSH, any configured **Ansible controllers**.

Each of those controllers will have inventories (lists) of end clients, which we will import to the DB to show and operate. The Uyuni Server will not contact those clients directly (at least as of the first implementation step) – only through their controllers.

We do not expect users to have more than let's say ~100 **controllers**, which would be a number that makes me start wondering about performance.

If ever we get close to that number, then requiring to register controllers as minions (and contacting them over the Salt bus instead of SSH) will make sense, and it is a step we are ready to make eventually.


Another way to look at this: the main initial intent here is to integrate with existing infrastructure, and the minimalistic/most universal way to do so is via SSH access. It's not equally nice to ask for controllers to be registered as minions just to get started. So we will start from SSH, and potentially offer Salt as an additional option at a later point, when it makes sense. In the meantime, nothing stops a user from registering a controller and manage it as usual.